src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/TagSet.cs

```diff
@@ -0,0 +1,235 @@
+/// Represents zero (null), one (string) or more (string[]) tags, avoiding the additional array overhead when necessary.
+/// </summary>
+[System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1066:Implement IEquatable when overriding Object.Equals", Justification = "Equals throws by intent")]
+internal readonly struct TagSet
+{
+    public static readonly TagSet Empty = default!;
+
+    private readonly object? _tagOrTags;
+
+    internal TagSet(string tag)
+    {
+        Validate(tag);
+        _tagOrTags = tag;
+    }
+
+    internal TagSet(string[] tags)
+    {
+        Debug.Assert(tags is { Length: > 1 }, "should be non-trivial array");
+        foreach (var tag in tags)
+        {
+            Validate(tag);
+        }
+
+        Array.Sort(tags, StringComparer.InvariantCulture);
+        _tagOrTags = tags;
+    }
+
+    public string GetSinglePrechecked() => (string)_tagOrTags!; // we expect this to fail if used on incorrect types
+    public Span<string> GetSpanPrechecked() => (string[])_tagOrTags!; // we expect this to fail if used on incorrect types
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1065:Do not raise exceptions in unexpected locations", Justification = "Intentional; should not be used")]
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Blocker Code Smell", "S3877:Exceptions should not be thrown from unexpected methods", Justification = "Intentional; should not be used")]
+    public override bool Equals(object? obj) => throw new NotSupportedException();
+
+    // [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1065:Do not raise exceptions in unexpected locations", Justification = "Intentional; should not be used")]
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Blocker Code Smell", "S3877:Exceptions should not be thrown from unexpected methods", Justification = "Intentional; should not be used")]
+    public override int GetHashCode() => throw new NotSupportedException();
+
+    public override string ToString() => _tagOrTags switch
+    {
+        string tag => tag,
+        string[] tags => string.Join(", ", tags),
+        _ => "(no tags)",
+    };
+
+    public bool IsEmpty => _tagOrTags is null;
+
+    public int Count => _tagOrTags switch
+    {
+        null => 0,
+        string => 1,
+        string[] arr => arr.Length,
+        _ => 0, // should never happen, but treat as empty
+    };
+
+    internal bool IsArray => _tagOrTags is string[];
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA2201:Do not raise reserved exception types", Justification = "This is the most appropriate exception here.")]
+    public string this[int index] => _tagOrTags switch
+    {
+        string tag when index == 0 => tag,
+        string[] tags => tags[index],
+        _ => throw new IndexOutOfRangeException(nameof(index)),
+    };
+
+    public void CopyTo(Span<string> target)
+    {
+        switch (_tagOrTags)
+        {
+            case string tag:
```

Comment:
no `null` case? (TagSet.Empty)

Reply:
there would be nothing to do - it would be `case null: break;`